### ISSUE TITLE
Improve error handling/logging for illegal_argument_exception|failed to build synonyms errors

### ISSUE THREAD

**Issue by ppf2** [elastic/elasticsearch#44243]
6.6.1, 6.7.0

This happens with the following synonym:

```
PUT test
{
  "settings": {
    "index": {
      "analysis": {
        "filter": {
          "synonyms_filter": {
            "type": "synonym_graph",
            "lenient": "true",
            "synonyms": [
                "アメリカン航空,aa,アメリカン"
              ],
            "tokenizer": "whitespace"
          }
        },
        "analyzer": {
          "japanese-search": {
            "filter": [
              "synonyms_filter"
            ],
            "type": "custom",
            "tokenizer": "custom_user_dictionary"
          }
        },
        "tokenizer": {
          "custom_user_dictionary": {
            "mode": "normal",
            "user_dictionary": "test-tokenization.txt",
            "type": "kuromoji_tokenizer",
            "discard_punctuation": "false"
          }
        }
      }
    }
  }
}
```

Because test-tokenization.txt contains this entry with unknown characters.

```
アメリカン航空,アメ��カン航空,アメリカンコウクウ,カスタム用語
```

As a result, the index cannot be created:

```
{
  "error": {
    "root_cause": [
      {
        "type": "illegal_argument_exception",
        "reason": "failed to build synonyms"
      }
    ],
    "type": "illegal_argument_exception",
    "reason": "failed to build synonyms",
    "caused_by": {
      "type": "array_index_out_of_bounds_exception",
      "reason": null
    }
  },
  "status": 400
}
```

Elasticsearch does not log anything to INFO/WARN/ERROR.  The full stack can only be found at DEBUG level.

```
[2019-07-11T11:30:48,629][DEBUG][o.e.c.m.MetaDataCreateIndexService] [node1] [test] failed to create
java.lang.IllegalArgumentException: failed to build synonyms
	at org.elasticsearch.index.analysis.SynonymTokenFilterFactory.buildSynonyms(SynonymTokenFilterFactory.java:158) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.analysis.SynonymGraphTokenFilterFactory.getChainAwareTokenFilterFactory(SynonymGraphTokenFilterFactory.java:51) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.analysis.CustomAnalyzerProvider.build(CustomAnalyzerProvider.java:87) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.analysis.AnalysisRegistry.processAnalyzerFactory(AnalysisRegistry.java:553) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.analysis.AnalysisRegistry.build(AnalysisRegistry.java:477) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.analysis.AnalysisRegistry.build(AnalysisRegistry.java:167) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.IndexService.<init>(IndexService.java:164) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.index.IndexModule.newIndexService(IndexModule.java:402) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.indices.IndicesService.createIndexService(IndicesService.java:526) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:480) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.metadata.MetaDataCreateIndexService$IndexCreationTask.execute(MetaDataCreateIndexService.java:469) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:643) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:270) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:200) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:135) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-6.7.0.jar:6.7.0]
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-6.7.0.jar:6.7.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_111]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_111]
Caused by: java.lang.ArrayIndexOutOfBoundsException
	at org.apache.lucene.analysis.tokenattributes.CharTermAttributeImpl.copyBuffer(CharTermAttributeImpl.java:44) ~[lucene-core-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:16:28]
	at org.apache.lucene.analysis.ja.JapaneseTokenizer.incrementToken(JapaneseTokenizer.java:486) ~[?:?]
	at org.apache.lucene.analysis.synonym.SynonymMap$Parser.analyze(SynonymMap.java:318) ~[lucene-analyzers-common-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:17:19]
	at org.elasticsearch.index.analysis.ESSolrSynonymParser.analyze(ESSolrSynonymParser.java:57) ~[elasticsearch-6.7.0.jar:6.7.0]
	at org.apache.lucene.analysis.synonym.SolrSynonymParser.addInternal(SolrSynonymParser.java:114) ~[lucene-analyzers-common-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:17:19]
	at org.apache.lucene.analysis.synonym.SolrSynonymParser.parse(SolrSynonymParser.java:70) ~[lucene-analyzers-common-7.7.0.jar:7.7.0 8c831daf4eb41153c25ddb152501ab5bae3ea3d5 - jimczi - 2019-02-04 23:17:19]
	at org.elasticsearch.index.analysis.SynonymTokenFilterFactory.buildSynonyms(SynonymTokenFilterFactory.java:154) ~[elasticsearch-6.7.0.jar:6.7.0]
	... 23 more
```

Suggestions:

1.  It can be helpful to move the exception stack to WARN or ERROR level since this represents a failure to create the index

2.  Including the actual synonym entry (entries) that are preventing the creation of the index in the response error message/stack will be helpful.  In this case, since it is an entry in the tokenization file that is causing an issue loading the synonym, it will also be helpful to reference the problem tokenization file name/path (and line number), etc..  It can be quite time consuming to manually go through hundreds of synonyms just to find the one/few that have the problem.  Certainly, it's easy after the fact (knowing that this particular issue is caused by invalid characters), but if you don't know the cause to begin with, having the actual failed entries/line numbers in the error logs will help quickly isolate the issue :)

**Comment by mocobeta** [comment]
@ppf2 
I'm not a member of Elasticsearch team, but once encountered similar (but not same) errors and just curious about the cause of the AIOOB. To me, the problem comes from JapaneseTokenizer - seems it does not properly handle the user dictionary you are using.

> Because test-tokenization.txt contains this entry with unknown characters.

Can you elaborate about "unknown characters" here? I am guessing they might be surrogate pair characters, is that correct?

**Comment by ppf2** [comment]
The Lucene [issue](https://issues.apache.org/jira/browse/LUCENE-8933) was [merged/backport](https://github.com/apache/lucene-solr/pull/810) to 8.3.0 as an improvement, i.e., available in Elasticsearch 7.5.0 and above.  Closing this issue at this time.  Thx!
